README.md
# Advanced Controllers

Features:

* MVC-like controllers for Express
* Easy configuration through `@Decorators`
* Data binding for request data (query, body, params)
* Can handle `async` functions and Promises
* Return value handling (data, exception, Promise)
* Authorization support (custom or roles and permissions)
* Written in TypeScript, compiled to ES5

See the tests for examples. See the [Change log](CHANGELOG.md) for breaking changes.


## Not documented yet

* Adding custom validators


## MVC Features

Inspired on ASP.NET MVC it is possible to create Express based controllers and actions.

```typescript
import * as web from 'advanced-controllers';

@web.Controller('/kitten')
class KittenController extends AdvancedController {
	// GET /kitten/all
	@web.Get('/all')
	getAll() { }

	// GET /kitten/details
	@web.Get()
	details() {}

	// POST /kitten/create
	@web.Post()
	create() {}

	// DELETE /kitten/delete
	@web.Del('delete')
	deleteKitten() {}
}

let kittenCtrl = new KittenController();
kittenCtrl.register(expressApp);
```

**Features:**

* Use the `@Controller('name')` decorator on the controller class
* Use the following functions: `Get`, `Post`, `Put`, `Head`, `Options`, `Del`
* The beginning `/` character in the names is optional
* The method names can be omitted. In this case the function name is used

**Caveats:**

* You have to actually *call* these functions. Good: `@Get()`, bad: `@Get`
* Use `@Del` instead of `@Delete`
* Don't forget to inherit from `AdvancedController` and call the `register()` function
* You forget to initialize `body-parser` for your Express app (so body won't be parsed)

### Registration details

You can register a single controller or all AdvancedController instances created so far.

```typescript
kittenCtrl.register(
	expressApp, // Express app
	{}, // Optional settings
);

// OR

AdvancedController.registerAll(expressApp, {});
```

Optional `settings` can have these props:

- `namespace`: Prefix for the routes (E.g. `namespace1` --> `/namespace1/ctrl/action`)
- `debugLogger`: Debug logger for registration
- `errorLogger`: Any runtime errors (E.g. Errors / promise rejects in actions -- except `WebError` instances, see below at the `WebError` section)
- `implicitPublic`: See below at Permissions

**Caveats:**

- If you use `AdvancedController.register`, use it once and do not use the instances' `register` methods.
- `AdvancedController.register` works on instantiated controllers.


## Data Binding

Tired of calling and validating `let myStuff = req.body.someVariable` in every function? Well, we try to make it a bit more comfortable. We have some limitations though but here's what we've got:

```typescript
import * as web from 'advanced-controllers';

@web.Controller('/kitten')
class KittenController extends web.AdvancedController {
	// GET /kittens/all?from=0[&cnt=25]
	@web.Get('/all')
	getKittens(
		@web.Query('from', Number) from: number,
		@web.Query('cnt', Number, true) cnt: number
	) { }

	// POST /kittens/create, body: { kitten: {} }
	@web.Post()
	create(
		@web.Body('kitten', Object) kitten: Kitten
	) {}

	// POST /kittens/create2, body: {}
	@web.Post()
	create2(@web.Body() kitten: Kitten)

	// DELETE /kittens/delete/<id>
	@web.Del('delete/:id')
	deleteKitten(@web.Param('id') id: string) {}
}
```

**Interface**

```typescript
// Whole body
export function Body(type?: any): ActionDecorator;
// Member of body object
export function Body(name: string, type?: any, optional?: boolean): ActionDecorator;
// Query winthout type
export function Query(name: string, optional?: boolean): ActionDecorator;
// Query with type
export function Query(name: string, type: any, optional?: boolean): ActionDecorator;
// Param with or without type
export function Param(name: string, type?: any): ActionDecorator;
```

**Features:**

* Type parameter is optional
  * In `Body`: only validation
  * In `Query` and `Param`: parse + validation
  * Supported types by default: `String`, `Number`, `Object`, `Array`
* If there is no type for `Query` and `Param` then the parameter value will be a `string`
* In `Query` and `Param`: Objects and arrays in query MUST be JSON-serialized. But seriously... arrays and objects in query?
* The bound value must be present unless you set the `optional` parameter to `true`


**Caveats:**

* Don't forget about the parentheses... Good: `@Body()`, bad: `@body`
* You MUST add the *variable name* when parsing body or query parameter. We cannot parse it for you
  * Well, ehm... actually we could (like the Angular team did) but currently we don't want to. It's kinda ugly. Maybe later


## Original Request and Response

You can access the original `req` or `res` objects with similar syntax. Beware: if you ask the `res` object then we won't handle the return values for you. (Return values are discussed soon.)

```typescript
@web.Controller('casual')
class CasualController extends web.AdvancedController {
	@web.Get('fancy-function')
	fancyFunction(
		@web.Req() req: web.Request,
		@web.Res() res: web.Response
	) {}
}
```

**Authenticated Users**

There's a `@Auth` shorthand which returns the `request.auth` object (or `undefined`). The value of the user is usually set in an express middleware such as `express-jwt`.

_Note: `express-jwt` versions below 7 used `request.user`, but from 7+ they use `request.auth`. The `@User` shorthand (deprecated) returns whichever it finds, the `@Auth` works with versions 7+._

```typescript
@web.Controller('casual2')
class CasualController extends web.AdvancedController {
	@web.Get('latest-fancy-function')
	latestFancyFunction(@web.Auth() auth?: { id: string }) {
		console.log(`Gotcha: ${user ? user.id : 'nevermind'}`);
	}

	// Deprecated
	@web.Get('another-fancy-function')
	anotherFancyFunction(@web.User() user?: { id: string }) {
		console.log(`Gotcha: ${user ? user.id : 'nevermind'}`);
	}
}
```


**Caveats:**

* If you use `res` then you have to manually end the request, e.g. `res.send('')` (see next section)
* Parentheses... Good: `@Req()`, bad: `@Req`
* The `@Auth` and the `@User()` decorators return `undefined` by default. You'll need an `express-jwt` to have anything there.


## Return values

By default the response is closed automatically with a status code and sometimes with data

* Async actions / Promise results are awaited first
* When the action executes correctly, depending on the *return value*
  * On `undefined`: 200
  * On `string` types: 200 + result as body (raw string)
  * Otherwise: 200 + value `JSON.strigify`-ed
* Missing bound parameter: 400
* When the action throws an error:
  * If the error has a `statusCode` property ending with `statusCode`
	* You should use the `WebError` class for this
  * Otherwise 500
  * If the error has a `json` field then it will be sent as JSON
  * If the error has a `text` field then it will be sent as plain text
  * Default error parsing error sends back `{ "errors": [ { "message": "some-stuff" }]}`

**Caveats:**

* If the action asked for `res` then there is no auto-close. In this case we don't know whether the response is closed -- or will be closed -- in the action. The only exception is if the action throws an error: in that case we apply the regular error handler logic.



## `WebError` object

This object extends `Error` and can be used to conveniently send back custom HTTP codes, error messages and codes. Feel free to throw it in actions, the framework will handle it.

* `new WebError(message: string)`, sending HTTP status code 500 by default
* `new WebError(message: string, statusCode: number)`
* `new WebError(message: string, settings: { statusCode?: number, errorCode?: number | string})`, the `errorCode` will be in the result JSON as `errors[0].code`

The result will be something like this: `{ "errors": [ { "message": "some-stuff", "code": 1 }]}` + HTTP 400 header

Customization by overwriting `WebError.requestErrorTransformer`.


## Middlewares

One extra functionality is the utilization of `express.use()`. You can specify middleware calls over the actions.

```typescript
@web.Controller('middleware')
class MiddlewareTestController extends web.AdvancedController {

	middleware(req: web.Request, res: web.Response, next: Function) {
		console.log('Middleware called');
		next();
	}

	@web.Get('do-stuff')
	@web.Middleware('middleware')
	doStuff() {}

	@web.Get('do-more-stuff')
	@web.Middleware(otherMiddlewareFunction)
	doMoreStuff() {}
}
```

**Functions:**

* Call a controller function -- function name (as string) in parameter
* Call directly a function -- function reference in parameter
* In both cases on middleware call the `this` reference will be the controller instance

**Limitations:**

* There's no way to add global middleware, won't be, do it manually
* There's no way to add class-specific middleware (this might change)
* There's no way to add independent middlewares with custom / RegEx routes, do it manually

**Caveats:**

* A tricky one: if you specify the middleware function with Arrow Syntax (`() => {}`) then the `this` reference won't refer to the controller instance when the middleware is called. The reason is TS/ES6 to ES5 transpilation: the `this` reference changes in the process and I could not bind it.


## Permissions

Action authorization can be controlled by the following decorators.
Controller classes and action functions can be decorated (the latter is stronger).
An action will have a single permission. (You shouldn't use multiple decorators on the same class or action function.)

- `@Permission(name?: string)`: The action requires the `name` permission (default value: `ctrl.action`)
- `@Authorize()`: The action requires an authenticated user, i.e. `req.user` object must not be `undefined`
- `@Public()`: The action does not require

The permission check can be managed 2 ways.

1. Default: custom permission check. Use a custom middleware before registering the controllers creating the following function on the request object: `req.user.hasPermission(permission: string): boolean | Promise<booleam>`. You can implement it however you'd like to
2. Role based: Call the `AdvancedController.setRoles(roles: { name: string, permissions: string[] }[])` function to set the roles and their permissions. The following array should exists: `req.user.roles: string[]`

```typescript
export interface RequestWithUser extends Request {
	user: {
		// Default: custom authorization
		hasPermission?(permission: string): boolean | Promise<boolean>;
		// Role-based authorization
		roles?: string[]
	};
}
```

You should create the `req.user.hasPermission` function OR the `req.user.roles` array in your custom authorization middleware.


### A security enforcement

If you don't use permissions (`@Permission` or `@Authorize` or `@Public`) then you can ignore this subsection.

If there are permission-related decorators in your app then you shall do at least one of the following to avoid auto-authentication:

- Decorate public functions (or controllers) with the `@Public` decorator
- Register the controllers with `implicitPublic`, e.g.: `AdvancedController.regiseterAll(app, { implicitPublic: true })`

This is to prevent unintentional publication of some of your actions by forgetting the `@Permission` decorator.


### Example

You can use permissions like this:

```typescript
// You can annotate this
@web.Controller('perm')
class PermissionController extends web.AdvancedController {

	// GET /perm
	// Needs permission: 'perm.testEmpty'
	@web.Permission()
	@web.Get('')
	testEmpty { return { done: true }; }

	// GET /perm/test1-a
	// Needs permission: 'perm.test1-a'
	@web.Permission()
	@web.Get('test1-a')
	testOneA() { return { done: true }; }

	// GET /perm/testOneB
	// Needs permission: 'perm.TestOneB'
	@web.Permission()
	@web.Get()
	testOneB() { return { done: true }; }

	// POST /perm/test2
	// Needs permission: 'perm.test-two'
	@web.Permission('perm.test-two')
	@web.Post('test2')
	testTwo() { return { done: true }; }

	// GET /perm/noPerm
	// NO permission required
	@web.Get('noperm')
	@web.Public()
	noPerm() { return { done: true }; }

	// GET /perm/authorized
	// Must be authenticated, but no explicit permission required
	@web.Get('authorized')
	@web.Authorize()
	authorized() { return { done: true}; }
}
```

### Static functions

The following static functions help the management of authorization:

- `AdvancedController.getAllPermissions(): string[]`: Aggregates the `getPermission()` results for all `AdvancedController` instances.
- `AdvancedController.getAllPublicRoutes(): string[]`: Aggregates the `getPublicRoutes()` results for all `AdvancedController` instances. This can be used to create a whitelist in he authentication middleware (e.g. skip JWT checks on these URLs). Note that the results contain `/ctrl/action` style URLs but they do NOT contain the `namespace` if there is any (i.e. NOT `/namespace/ctrl/action`).

Note that the static functions work on instantiated controllers only.


### Notes and Caveats

Notes:

* Note that this package authorizes but NOT authenticates.
* Note that only one permission check method works at a time. By default: `hasPermission`. After `AdvancedController.setRoles` is called: role-based.
* The permission-related decorator can be used on classes as well. It won't override the action-level permissions though.
* When the user is not authenticated (most commonly: `req.user` does not exists) the response is: 401 `{ errors: [{ message: "Unauthenticated" }]}`.
* When the user is authenticated but does not have the required permissions the response is: 403 `{ errors: [{ message: "Unauthorized" }]}`.

See [HTTP Status Codes Wiki][https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#4xx_Client_Error].

Caveats:

* Register the middleware providing `req.user.hasPermission` or `req.user.roles` *before* registering the controller.
* On empty action name: the permission will have the name of the function instead of the name of the action (which is `''`)


CHANGELOG.md
# CHANGELOG

## 2.5.0 (2023-05-04)

- Introduce `@Auth()` and deprecate `@User()` to support `express-jwt` version 7 and above.

## 2.4.1 (2023-01-02)

- Update deps
- Fix breaking tests


## 2.4.0 (2021-03-04)

- Better error logging: besides the error object, send a lot of useful info to the logger.
- Update deps


## 2.3.0 (2018-11-17)

- Added `@User()` attribute decorator, shorthand for `request.user`, useful for authorization.
- Changelog styling updated
- Update deps


## 2.2.3 (2018-08-23)

- Validator for booleans -- Contribution from [kundralaci](https://github.com/kundralaci)
- Test for boolean bindings
- Update deps


## 2.2.2 (2018-07-07)

- Update dependencies


## 2.2.1 (2018-01-18)

**Features**

- `WebError` can have text `errorCode` values as well
- Update dependencies


## 2.2.0 (2017-11-02)

**Fixes**

- Add `.js.map` files to npm package
- Add comments (JSDoc) to npm package
- Close rejected async requests even if autoClose is set (`Response` object is used)
- Actions with the same route and different method won't have their middlewares interfere with each other
	- Now a middleware is applied if the route AND the method are ok. The latter is the news.
- Middlewares are applied in the same order as their declaration
- Readme fixes and update
- Test fixes

**Changes**

- Results of `string` or a `Promise<string>` wont't be JSON-serialized
	- Technically it's a breaking change but sending JSON-serialized strings is clearly a bad practice noone should use
- Error callback changed as it was already documented
	- Debug callback already works this way


## 2.1.1 (2017-07-10)

**Fixes**

- Permission name generation for actions with routes `/`, `/:*` would also generate `ctrlname:funcname`


## 2.1.0 (2017-07-10)

**Changes** (undocumented breaking)

- `@Get('')` and other methods will register to `/ctrlname` instead of `/ctrlname/`
- Permission for `Get('')` will be `ctrlName.funcName`
- Using different permissions on same route + different method is available. Using any permission and no permission (Public) on the same route + different method is still forbidden.


## 2.0.0 (2017-06-28)

**Fixes**

- Fixes in readme (typos, errors)


**BREAKING CHANGES**: Renames

- `@AllowAnonymus` --> `@Public`
- `getAllWhiteList` --> `getAllPublicRoutes`
- `getWhiteList` --> `getPublicRoutes`


## 1.1.0 (2017-06-28)

**Features**

- Less constraints on `Body`, `Query` and `Param`: each can be used without types
- Minor undocumented change: `@Body()` won't validate `body` to be an Object


## 1.0.0 (2017-06-28)

**Features**

- Introduced some static messages on `AdvancedController` affecting every existing controller instances
- Role-based authentication + authorization
- `Param` decorator for routing params (e.g. `my-ctrl/action/:id`)
- `AllowAnonymus` and `Authorize` decorators
- WhiteLists
- Some checks to prevent various unintentional cases (e.g. 2 controllers using the same name)

**BREAKING CHANGES**

- Rename all decorators: `camelCase` --> `PascalCase` (e.g. `get` --> `Get`)
- Type rename: `Req` --> `Request`, `Res` --> `Response`
- Rename `BaseController` --> `AdvancedController`
- Rename at `AdvancedController` instance: `getAllPermissions` --> `getPermissions`
- Default permission goes from `<ctrl>:<action>` to `<ctrl>.<action>`
- TS build: Requires full ES2015 lib (specifically: `ES2015.collections`, even more specifically: `Map`)


3# 0.4.0 (2017-05-05)

**Features**

- Permission support on classes


## 0.3.2 (2017-05-04)

**Fixes**

* Fix enables using controller classes having a member variable with `function` type


## 0.3.1 (2017-02-23)

**Features**

* `WebError` can be used to send back a `code` besides a message. `{ errors: [ { message: string, code?: number }] }`

No breaking change, only extension.


## 0.3.0 (2017-01-11)

**Features**

* Permission support
* Namespace support

**BREAKING CHANGES**

* Obsolete decorators like `@queryString` and `@bodyNumber` are removed

**Notes**

* Code is restructured into several files
* Removed `typings`, now using `@types/<name>` NPM packages
* `Promise` objects were already required; now TS build explicitly requires `es2015.promise`
* Minor git and package config changes
* Planning to go to `es2015` package build


## 0.2.0 (2016-04-17)

**Features**

* Can parse whole body (#2)
* Custom validation and parsing can be added
* Error handling is more customizable: can send back plain text or JSON object
  * Default settings: sending back something like [JSON API](http://jsonapi.org/format/): `"{"errors":[{"message":"<ERROR-MSG>"}]}"`
* `@queryTYPE` and `@bodyTYPE` are obsolete; use `@query(name, type, opt?)` instead

**BREAKING CHANGES**

* Error messages are not plain text but JSON objects by default


lib/index.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var _ = require('lodash');
function controller(controllerName) {
    if (controllerName[0] !== '/')
        controllerName = '/' + controllerName;
    return function (target) {
        target.__controller = { name: controllerName };
    };
}
exports.controller = controller;
function getHttpAction(prop) {
    if (_.isFunction(prop) && prop.action) {
        return prop.action;
    }
    return null;
}
function setHttpAction(prop, method, url) {
    if (url[0] !== '/')
        url = '/' + url;
    prop.action = prop.action || {};
    prop.action.url = url;
    prop.action.method = method;
    prop.action.params = prop.action.params || [];
    prop.action.middlewares = prop.action.middlewares || [];
}
function routeDeclaration(method, name) {
    return function (target, key, value) {
        setHttpAction(target[key], method, name || key);
    };
}
function get(name) { return routeDeclaration('GET', name); }
exports.get = get;
function post(name) { return routeDeclaration('POST', name); }
exports.post = post;
function put(name) { return routeDeclaration('PUT', name); }
exports.put = put;
function head(name) { return routeDeclaration('HEAD', name); }
exports.head = head;
function options(name) { return routeDeclaration('OPTIONS', name); }
exports.options = options;
function del(name) { return routeDeclaration('DELETE', name); }
exports.del = del;
function addMiddleware(prop, mwFunc) {
    prop.action = prop.action || {
        url: null,
        method: null,
        params: [],
        middlewares: [],
    };
    prop.action.middlewares.push(mwFunc);
}
function middleware(middlewareFunc) {
    return function (target, funcName) {
        if (typeof middlewareFunc === 'string') {
            addMiddleware(target[funcName], target[middlewareFunc]);
        }
        else {
            addMiddleware(target[funcName], middlewareFunc);
        }
    };
}
exports.middleware = middleware;
function addParam(prop, name, index, from, type, opt) {
    prop.action = prop.action || {
        url: null,
        method: null,
        params: [],
        middlewares: [],
    };
    prop.action.params.push({ index: index, from: from, name: name, type: type, opt: opt });
}
function req() {
    return function (target, propertyKey, parameterIndex) {
        var prop = target[propertyKey];
        addParam(prop, null, parameterIndex, 'req', null, false);
    };
}
exports.req = req;
function res() {
    return function (target, propertyKey, parameterIndex) {
        var prop = target[propertyKey];
        addParam(prop, null, parameterIndex, 'res', null, false);
    };
}
exports.res = res;
function addParamBinding(name, optional, from, type) {
    return function (target, propertyKey, parameterIndex) {
        var prop = target[propertyKey];
        addParam(prop, name, parameterIndex, from, type, optional);
    };
}
function obsWarn(oldAnnotation, newAnnotation) {
    console.warn("Obsolete " + oldAnnotation + ", use " + newAnnotation + " instead");
}
function bodyString(name, optional) { obsWarn('bodyString', 'body'); return addParamBinding(name, optional, 'body', String); }
exports.bodyString = bodyString;
function bodyNumber(name, optional) { obsWarn('bodyNumber', 'body'); return addParamBinding(name, optional, 'body', Number); }
exports.bodyNumber = bodyNumber;
function bodyObject(name, optional) { obsWarn('bodyObject', 'body'); return addParamBinding(name, optional, 'body', Object); }
exports.bodyObject = bodyObject;
function bodyArray(name, optional) { obsWarn('bodyArray', 'body'); return addParamBinding(name, optional, 'body', Array); }
exports.bodyArray = bodyArray;
exports.body = function (nameOrType, type, optional) {
    if (typeof nameOrType === 'string') {
        return addParamBinding(nameOrType, optional, 'body', type);
    }
    return addParamBinding(null, false, 'full-body', nameOrType || Object);
};
function queryString(name, optional) { obsWarn('queryString', 'query'); return addParamBinding(name, optional, 'query', String); }
exports.queryString = queryString;
function queryNumber(name, optional) { obsWarn('queryNumber', 'query'); return addParamBinding(name, optional, 'query', Number); }
exports.queryNumber = queryNumber;
function queryObject(name, optional) { obsWarn('queryObject', 'query'); return addParamBinding(name, optional, 'query', Object); }
exports.queryObject = queryObject;
function queryArray(name, optional) { obsWarn('queryArray', 'query'); return addParamBinding(name, optional, 'query', Array); }
exports.queryArray = queryArray;
function query(name, type, optional) { return addParamBinding(name, optional, 'query', type); }
exports.query = query;
var validators = [
    { type: String, check: _.isString, parse: function (input) { return input; } },
    { type: Number, check: _.isNumber, parse: parseInt },
    { type: Object, check: _.isObject, parse: JSON.parse },
    { type: Array, check: _.isArray, parse: JSON.parse },
];
function addValidator(validator) {
    if (_.some(validators, { type: validator.type })) {
        throw new Error("Cannot add validator with type " + validator.type + ": already parsing that!");
    }
    validators.push(validator);
}
exports.addValidator = addValidator;
var WebError = (function (_super) {
    __extends(WebError, _super);
    function WebError(message, statusCode) {
        if (statusCode === void 0) { statusCode = 500; }
        _super.call(this, message);
        this.statusCode = statusCode;
        WebError.requestErrorTransformer(this, message, statusCode);
    }
    WebError.requestErrorTransformer = function (error, message, statusCode) {
        error.json = { errors: [{ message: message }] };
    };
    return WebError;
}(Error));
exports.WebError = WebError;
function handleError(err, res) {
    err.statusCode = err.statusCode || 500;
    if (err.json) {
        return res.status(err.statusCode).json(err.json);
    }
    if (err.text) {
        return res.status(err.statusCode).send(err.text);
    }
    return res.sendStatus(err.statusCode);
}
function registerControllerFunction(thisBind, app, actionFunc, logger) {
    var action = getHttpAction(actionFunc);
    if (!action) {
        return false;
    }
    if (!action.method || !action.url) {
        throw new Error('Action has no method: ' + actionFunc);
    }
    var controllerName = thisBind.constructor.__controller.name;
    var url = controllerName + action.url;
    logger && logger('debug', "Registering " + action.method + " " + url + " [" + action.params.map(function (p) { return p.name; }) + "]");
    for (var _i = 0, _a = action.middlewares; _i < _a.length; _i++) {
        var mwFunc = _a[_i];
        logger && logger('debug', "Registering " + action.method + " " + url + " *MW*");
        app.use(url, mwFunc.bind(thisBind));
    }
    var binders = [];
    var autoClose = true;
    var _loop_1 = function(bind) {
        if (bind.from === 'req') {
            binders.push(function (params, req, res) { params[bind.index] = req; });
            return "continue";
        }
        if (bind.from === 'res') {
            binders.push(function (params, req, res) { params[bind.index] = res; });
            autoClose = false;
            return "continue";
        }
        var validator = validators.filter(function (item) { return item.type === bind.type; })[0];
        if (!validator)
            throw new Error("No validator for type: " + bind.type);
        if (bind.from === 'full-body') {
            binders.push(function (params, req, res) {
                if (req.body === undefined) {
                    throw new WebError("Empty Body", 400);
                }
                if (!validator.check(req.body))
                    throw new Error("Invalid value: Body should be a " + bind.type);
                params[bind.index] = req.body;
            });
            return "continue";
        }
        if (bind.from === 'body') {
            binders.push(function (params, req, res) {
                var parsed = req.body[bind.name];
                if (parsed === undefined) {
                    if (!bind.opt)
                        throw new WebError("Missing property: " + bind.name, 400);
                    params[bind.index] = undefined;
                    return;
                }
                if (!validator.check(parsed))
                    throw new Error("Invalid value: " + bind.name + " should be a " + bind.type);
                params[bind.index] = parsed;
            });
            return "continue";
        }
        if (bind.from === 'query') {
            if (!validator.parse)
                throw new Error("No parser in validator for type: " + bind.type + "; required when binding to query params");
            binders.push(function (params, req, res) {
                var value = req.query[bind.name];
                if (value === undefined) {
                    if (!bind.opt)
                        throw new WebError("Missing property: " + bind.name, 400);
                    params[bind.index] = undefined;
                    return;
                }
                var parsed = validator.parse(value);
                if (!validator.check(parsed))
                    throw new Error("Invalid value: " + bind.name + " should be a " + bind.type);
                params[bind.index] = parsed;
            });
            return "continue";
        }
    };
    for (var _b = 0, _c = action.params; _b < _c.length; _b++) {
        var bind = _c[_b];
        _loop_1(bind);
    }
    var actionProcessor = function (req, res) {
        var params = new Array(action.params.length);
        try {
            for (var _i = 0, binders_1 = binders; _i < binders_1.length; _i++) {
                var binder = binders_1[_i];
                binder(params, req, res);
            }
            var result_1 = actionFunc.apply(thisBind, params);
            if (!autoClose)
                return;
            if (result_1 === undefined)
                return res.sendStatus(200);
            else if (result_1 instanceof Promise) {
                result_1
                    .then(function (response) { return (result_1 !== undefined) ? res.json(response) : res.sendStatus(200); })
                    .catch(function (ex) {
                    (!ex.statusCode) && logger && logger('error', 'Something broke (Promise)', { ex: ex, message: ex.message, stack: ex.stack });
                    handleError(ex, res);
                });
            }
            else {
                res.json(result_1);
            }
        }
        catch (ex) {
            (!ex.statusCode) && logger && logger('error', 'Something broke (Exception)', { ex: ex, message: ex.message, stack: ex.stack });
            handleError(ex, res);
        }
    };
    app[_.toLower(action.method)](url, actionProcessor);
}
function getAllFuncs(obj) {
    var props = [];
    var protoObj = obj;
    while (protoObj) {
        props = props.concat(Object.getOwnPropertyNames(protoObj));
        protoObj = Object.getPrototypeOf(protoObj);
    }
    return props.sort().filter(function (e, i, arr) {
        if (e !== arr[i + 1] && typeof obj[e] === 'function')
            return true;
    });
}
function registerController(controller, app, logger) {
    var ctor = controller.constructor;
    if (!ctor || !ctor.__controller || !ctor.__controller.name) {
        throw new Error('Must use @controller decoration on controller!');
    }
    var funcNames = getAllFuncs(controller);
    for (var _i = 0, funcNames_1 = funcNames; _i < funcNames_1.length; _i++) {
        var name_1 = funcNames_1[_i];
        var action = ctor.prototype[name_1];
        if (getHttpAction(action) !== null) {
            registerControllerFunction(controller, app, action, logger);
        }
    }
}
var BaseController = (function () {
    function BaseController() {
    }
    BaseController.prototype.register = function (app, logger) {
        if (logger === void 0) { logger = console.log.bind(console); }
        registerController(this, app, logger);
    };
    return BaseController;
}());
exports.BaseController = BaseController;
//# sourceMappingURL=index.js.map